src/python-bindings/htcondor2/remote_param.cpp

```diff
@@ -0,0 +1,245 @@
+        PyErr_SetString( PyExc_IOError, "Failed to send request for parameter names." );
+        return NULL;
+    }
+    if(! sock.end_of_message()) {
+        // This was HTCondorIOError in version 1.
+        PyErr_SetString( PyExc_IOError, "Failed to send EOM for parameter names." );
+        return NULL;
+    }
+
+    sock.decode();
+
+    std::string reply;
+    if(! sock.code(reply)) {
+        // This was HTCondorIOError in version 1.
+        PyErr_SetString( PyExc_IOError, "Failed to receive reply for parameter names." );
+        return NULL;
+    }
+
+    if( reply == "Not defined" ) {
+        if(! sock.end_of_message()) {
+            // This was HTCondorIOError in version 1.
+            PyErr_SetString( PyExc_IOError, "Failed to receive EOM from remote daemon (unsupported version)." );
+            return NULL;
+        }
+
+        // Strictly speaking, this could be a reply from a daemon earlier
+        // than version 8.1.2, but let's not add another round-trip to
+        // half-ass a check by looking for a known knob name.
+
+        // This was HTCondorReplyError in version 1.
+        PyErr_SetString( PyExc_IOError, "Not authorized to query remote daemon." );
+        return NULL;
+    }
+    if( reply[0] == '!' ) {
+        sock.end_of_message();
+
+        // This was HTCondorIOError in version 1.
+        PyErr_SetString( PyExc_IOError, "Remote daemon failed to get parameter names." );
+        return NULL;
+    }
+
+    std::vector<std::string> keys;
+    if(! reply.empty()) {
+        keys.push_back(reply.c_str());
+    }
+
+    std::string key;
+    while(! sock.peek_end_of_message()) {
+        if(! sock.code(key)) {
+            // This was HTCondorIOError in version 1.
+            PyErr_SetString( PyExc_IOError, "Failed to read parameter name." );
+            return NULL;
+        }
+
+        keys.push_back(key.c_str());
+    }
+
+    if(! sock.end_of_message()) {
+        // This was HTCondorIOError in version 1.
+        PyErr_SetString( PyExc_IOError, "Failed to receive final EOM for parameter names." );
+        return NULL;
+    }
+
+    return PyUnicode_FromString(join(keys, ",").c_str());
+}
+
+
+static PyObject *
+_remote_param_get( PyObject *, PyObject * args ) {
+    // _param_get(self.location._handle, key)
+    PyObject_Handle * handle = NULL;
+    const char * key = NULL;
+
+    if(! PyArg_ParseTuple( args, "Os", (PyObject **)& handle, & key )) {
+        // PyArg_ParseTuple() has already set an exception for us.
+        return NULL;
+    }
+
+    auto * location = (ClassAd * )handle->t;
+
+    ReliSock sock;
+    bool result = start_config_command( CONFIG_VAL, sock, * location );
+    if(! result) {
+        // start_config_command() has already set an exception for us.
+        return NULL;
+    }
+
+
+    sock.encode();
+
+    if(! sock.put(key)) {
+        // This was HTCondorIOError in version 1.
+        PyErr_SetString( PyExc_IOError, "Can't send requested param name." );
```

Comment:
``Can't`` or ``Failed to``? Likewise for line 160

Reply:
That string is unmodified from the corresponding exception in version 1.  I didn't think the difference between "failed to" and "can't" in this context was significant enough (or confusing enough) in this context to warrant a change, but I guess "Failed to" would also be more consistent.  I can make the change if you want.

Reply:
I think it is fine then if you are preserving the old messages